rclcpp/test/rclcpp/test_graph_listener.cpp

```diff
@@ -0,0 +1,309 @@
+constexpr char node_namespace[] = "ns";
+constexpr char shutdown_error_str[] = "GraphListener already shutdown";
+
+}  // namespace
+
+class TestGraphListener : public ::testing::Test
+{
+public:
+  void SetUp()
+  {
+    rclcpp::init(0, nullptr);
+    node_ = std::make_shared<rclcpp::Node>(node_name, node_namespace);
+
+    // This dynamic cast is not necessary for the unittests, but instead is used to ensure
+    // the proper type is being tested and covered.
+    node_graph_ =
+      dynamic_cast<rclcpp::node_interfaces::NodeGraph *>(node_->get_node_graph_interface().get());
+    ASSERT_NE(nullptr, node_graph_);
+
+    using rclcpp::contexts::get_global_default_context;
+    graph_listener_ =
+      std::make_shared<rclcpp::graph_listener::GraphListener>(get_global_default_context());
+  }
+
+  void TearDown()
+  {
+    rclcpp::shutdown();
+  }
+
+protected:
+  std::shared_ptr<rclcpp::Node> node() {return node_;}
+  const rclcpp::node_interfaces::NodeGraph * node_graph_const() const {return node_graph_;}
+  rclcpp::node_interfaces::NodeGraph * node_graph() {return node_graph_;}
+  std::shared_ptr<rclcpp::graph_listener::GraphListener> graph_listener() {return graph_listener_;}
+
+private:
+  std::shared_ptr<rclcpp::Node> node_;
+  rclcpp::node_interfaces::NodeGraph * node_graph_;
+  std::shared_ptr<rclcpp::graph_listener::GraphListener> graph_listener_;
+};
+
+/* Run base class init/shutdown */
+TEST_F(TestGraphListener, construction_and_destruction) {
+  void();
+}
+
+// Required for mocking_utils below
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, !=)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, >)
+
+/* Error creating a new graph listener */
+TEST_F(TestGraphListener, error_construct_graph_listener) {
+  using rclcpp::contexts::get_global_default_context;
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_guard_condition_init, RCL_RET_ERROR);
+
+  using rclcpp::contexts::get_global_default_context;
+  RCLCPP_EXPECT_THROW_EQ(
+  {
+    auto graph_listener_error =
+    std::make_shared<rclcpp::graph_listener::GraphListener>(get_global_default_context());
+  }, std::runtime_error("failed to create interrupt guard condition: error not set"));
+}
+
+// Required for mocking_utils below
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, !=)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, >)
+
+/* Errors that occur when initializing the graph_listener */
+TEST_F(TestGraphListener, error_start_graph_listener) {
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_wait_set_init, RCL_RET_ERROR);
+    RCLCPP_EXPECT_THROW_EQ(
```

Comment:
For this macro, you can use it like `EXPECT_THROW` and you only need curly braces if you're testing multiple lines of code. Fix here and below

Reply:
See f13344a